ingestion/src/metadata/profiler/interface/sqlalchemy/profiler_interface.py

```diff
@@ -530,8 +530,15 @@ def fetch_sample_data(self, table, columns) -> TableData:
         sampler = self._get_sampler(
             table=table,
         )
-
-        return sampler.fetch_sample_data(columns)
+        sample = sampler.fetch_sample_data(columns)
+        # TODO: this is innefficient, we should change the schema to be a string and
+        # manage the casting when creating the TableData object with the logic coupled to OpenMetdata
+        # DataType / Pydantic instead of letting the underlying ORM / DB client handle it.
+        for i, row in enumerate(sample.rows):
+            sample.rows[i] = list(row)
+            for j, value in enumerate(row):
+                sample.rows[i][j] = self._json_safe(value)
+        return sample
```

Comment:
@sushi30 what case are we handling in this scenario?

Reply:
```
[2024-07-09 12:56:10] DEBUG    {metadata.OMetaAPI:table_mixin:76} - Traceback (most recent call last):
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/ingestion/ometa/mixins/table_mixin.py", line 73, in ingest_table_sample_data
    data=sample_data.model_dump_json(),
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/ingestion/models/custom_pydantic.py", line 65, in model_dump_json
    self.model_dump(
  File "/Users/imri/projects/OpenMetadata-pycharm/.venv/lib/python3.11/site-packages/pydantic/main.py", line 347, in model_dump
    return self.__pydantic_serializer__.to_python(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf7 in position 10: invalid utf-8
```

also added to the PR description

Reply:
Do you know the type of the column and for what service it happens. Basically, [here](https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/src/metadata/profiler/orm/types/custom_hex_byte_string.py) we convert `BYTES` column type result on SQA side to a string representation of the bytes. 

We might be missing this for a specific type/service which might cause the `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf7 in position 10: invalid utf-8` error.

Reply:
If I understand correctly, this requires defining a custom type map per database:
https://github.com/open-metadata/OpenMetadata/blob/3c36a3f1b27e52c2cb7785156d4feb189e7865b0/ingestion/src/metadata/profiler/orm/converter/snowflake/converter.py#L28

This will not scale as we add more connectors. We need something to catch this dynamically.

I can introduce a patch for mssql specifically but this fix will still be required regardless to avoid introducing un-handeled data types in the future. With current implementation, any introduction of a new data source or new data type can cause the sample ingestion to fail and there is no way of detecting it until its in service.

Reply:
It is not per service, but by type, but some services have their own specific type (hence the question around the service). In this case, we would create the custom type to handle the specific type that returns a raw bytes string. 

We should handle that decode error at the source using the SQA Custom type. This patch has O(n^2) efficiency so it can quickly become quite slow.

Reply:
Because each service can introduce its own arbitrary types, in the end it will be defined by service with some potential to reuse logic between services.

In this example, Snowflake defines a type of `BINARY` and mssql uses something like `BYTES` but another service might define `BYTESBUFFER` and so on.

The root issue here is that `TableData.rows` can contain any type of object so there is no way of guaranteeing that the [serializer running the the sink](https://github.com/open-metadata/OpenMetadata/blob/b960b60965593c584247f8bba3506295cc000377/ingestion/src/metadata/ingestion/ometa/mixins/tests_mixin.py#L344) will know how to handle it.

I can look at defining this at the serializer level in pydantic to replace the value with something like `<UNSERIALIZABLE_VALUE>`. This can be done by implementing a [CustomSerializer](https://docs.pydantic.dev/latest/concepts/serialization/#dictmodel-and-iteration) for the `TableData` type.

How does that sound?

Reply:
> Because each service can introduce its own arbitrary types
That is true at the marge. `BINARY` and `BYTES` are used across many different services. Creating a custom type to handle binary will then be used across all services. 

My point is that we should only serialize what we need to serialize and where we need to.

Reply:
Added handling using a custom pydantic serializer for `TableData`